Cinematic spectator camera

eps0003 opened the pull request:
## Status

**IN DEVELOPMENT**

## Description

This PR greatly improves the viewing experience of spectators by adding a cinematic camera mode. You can enable the cinematic camera by right-clicking while in spectator or while dead/respawning. Moving the camera using mouse/WASD keys or clicking a player to follow them disables the cinematic camera. The cinematic camera is enabled by default.

The cinematic camera works by positioning the camera at the mean position of all players. The zoom level is the max distance between the two furthest players on the horizontal and vertical axes (clamped between min and max default zoom). When a round starts, round ends, during build time or when there are not enough players for a match, the camera zooms out to fit the entire map (or as much as it can zoom out to fit).

The cinematic camera also focuses on important events. The events are listed below, ordered by importance:

1. Missing flags
2. Capturing halls
3. Builders near enemy flag
4. Players near enemy tent
5. Lit kegs
6. Capturing vehicles with spawns (ballistae, war boats)

If there are multiple important events, the events with higher importance are prioritised. Kegs closer to exploding and vehicles closer to being captured are more important than other kegs/vehicles.

This PR also improves camera panning while the 'snap camera to pixel' setting is enabled.

## Steps to Test or Reproduce

1. Join any game mode with spectator team (event focus works best on CTF/TTH)
2. Join spectator
3. Right click if the camera isn't already cinematic

This has been thoroughly tested on Gather and in TDM tournaments.

Harrison-Miller commented:
Do you have any cool gifs or videos displaying this working?

eps0003 commented:
I've only got outdated clips from TDM tournaments which only show the camera pan/zoom. I might make a few gifs soon. https://www.twitch.tv/eps0003/clips?filter=clips&range=all

Harrison-Miller commented on Rules/CommonScripts/Spectator.as:
```diff
@@ -15,101 +27,188 @@ CPlayer@ targetPlayer()
+				posTarget += blob.getInterpolatedPosition();
+
+				//look for largest distance between two players
+				for (uint j = 0; j < players.length; j++)
```
start at i+1

Harrison-Miller replied:
and the end condition above that should be player.lenght-1

eps0003 replied:
I can't do `player.length - 1` because the same loop is used for calculating the average position of all players. This would skip over the last player and make the camera sit off centre.

Harrison-Miller replied:
It's not an important enough improvement to worry too much about it just leave` i < player.length`. You can still do j = i+1. (If you really wanted you can do `posTarget += players.last().getInterpolatedPosition()` before calculating the average.

Harrison-Miller commented on Rules/CommonScripts/Spectator.as:
```diff
@@ -15,101 +27,188 @@ CPlayer@ targetPlayer()
+					u32 distX = Maths::Abs(blob.getPosition().x - blob2.getPosition().x);
+					u32 distY = Maths::Abs(blob.getPosition().y - blob2.getPosition().y);
+					maxDistX = Maths::Max(distX, maxDistX);
+					maxDistY = Maths::Max(distY, maxDistY);
```
maybe consider an early out if max distance reached, I'm interested to see the performance of this running every tick on a full server.

Harrison-Miller commented on Rules/CommonScripts/Spectator.as:
```diff
@@ -15,101 +27,188 @@ CPlayer@ targetPlayer()
+		}
+
+		//adjust camera pos and zoom
+		camera.targetDistance += (zoomTarget - camera.targetDistance) / CINEMATIC_ZOOM_EASE * getRenderApproximateCorrectionFactor();
```
if it turns out that this is really slow on large servers, you could simply calculate the focus/targetPos every few ticks, but update the easing every tick.

Harrison-Miller commented on Rules/CommonScripts/PlayerCamera.as:
```diff
@@ -205,3 +211,32 @@ void onRender(CRules@ this)
+		if (controls.isKeyJustPressed(KEY_RBUTTON) && (spectatorTeam || diffTime <= 0))
+		{
+			SetTargetPlayer(null);
+			timeToCinematic = 0.0f;
```
why are you automatically sending the player into cinematic mode everywhere in this file. At the very least there is a bug with the death lock which zooms in. When you set timeToCinematic = 0.0f; deathlock won't zoom in on the locked position because of the condition on line 128 of Spectator.as.

Harrison-Miller replied:
ignore everything in this comment, rereading I was wrong about the bug. I do think however what you talked about was it not auto going to cinematic on some of those conditions but instead remembering.

eps0003 replied:
I don't understand what you mean. Death lock should still work like normal. Look at `PlayerCamera.as` from around line 135.

eps0003 commented on Rules/CommonScripts/Spectator.as:
```diff
@@ -1,7 +1,19 @@
+const u8 CINEMATIC_ZOOM_EASE = 14.0f;					//amount of ease when zooming while cinematic
+const u8 CINEMATIC_PAN_X_EASE = 12.0f;					//amount of ease along the y-axis while cinematic
+const u8 CINEMATIC_PAN_Y_EASE = 12.0f;					//amount of ease along the x-axis while cinematic
+const u32 CINEMATIC_TIME = 10.0f * getTicksASecond();	//time until camera automatically becomes cinematic in seconds
```
That's good to know

eps0003 commented:
I've made the changes @Harrison-Miller suggested and made it remember whether the cinematic camera is enabled or disabled (suggested by @furai). This should be ready after a final review.

1bardesign commented:
Will wait for verra to re-review once he's back but thanks for making the requested changes quickly!

The performance concerns mean that it'd be great to have a big testing session (or at least a lot of bots added), but I guess it's at least only clientside performance issues.

eps0003 commented:
This is really laggy on a CTF server with only 12 players. I'll implement the performance improvements soon.

eps0003 commented:
Even with the performance improvements suggested by @Harrison-Miller, this is still extremely laggy. Increasing the update interval will make the camera stop and start as it pans/zooms. I don't really know what else could be done to improve performance. 😢

1bardesign commented:
I feel like updating/selecting points of interest iteratively (ie over several frames, process 1 blob at a time and then sort generated points), and then just panning there as a separate process would probably be the way forward. Likely to require an overhaul of the code though.

eps0003 commented:
I'm glad I took a break from this because I now see so many issues with how I implemented it. Here's what I have changed to improve the performance of this:
- Moved most logic out of onRender
- Filtered out unimportant blobs so sorting is much quicker
- Implemented quicksort to replace the shitty bubble sort algorithm
- Replaced the code to position and scale the camera with more efficient code
  - I used to compare every player blob with every other player blob
    `n * n/2` steps
  - Now I quicksort twice to get the furthest distance between two player blobs on each axis and then loop through all player blobs to calculate the mean position
    `2*log(n) + n` steps

This should now work with good framerate on populated servers
<details>
<summary>Graph</summary>

![image](https://user-images.githubusercontent.com/30195802/68985926-5bee7380-086e-11ea-8fe7-0c0152cba416.png)
</details>

eps0003 commented:
Optimised it even further. Now it takes `n` steps to calculate the position and scale of the camera.

Harrison-Miller commented on Rules/CommonScripts/PlayerCamera.as:
```diff
@@ -197,11 +203,94 @@ void onRender(CRules@ this)
+		if (getGameTime() == 1)
+		{
+			//initially position camera to view entire map
+			CCamera@ camera = getCamera();
```
Replace with the ViewEntireMap function in CinematicCommon.as

Harrison-Miller replied:
Replace with ViewEntireMap function.

Harrison-Miller commented on Rules/CommonScripts/PlayerCamera.as:
```diff
@@ -197,11 +203,94 @@ void onRender(CRules@ this)
+			camera.targetDistance = Maths::Min(zoomW, zoomH);
+		}
+
+		if (this.isMatchRunning() && !this.isWarmup() && !this.isGameOver())
```
only need this.isMatchRunning

Harrison-Miller replied:
Only need this.isMatchRunning()

Harrison-Miller commented on Rules/CommonScripts/PlayerCamera.as:
```diff
@@ -197,11 +203,94 @@ void onRender(CRules@ this)
+						Vec2f pos = blob.getPosition();
+
+						//max distance along each axis
+						if (pos.x > maxPos.x) maxPos.x = pos.x;
```
can use min/max or clamp functions.

Harrison-Miller commented on Rules/CommonScripts/CinematicCommon.as:
```diff
@@ -0,0 +1,303 @@
+	zoomTarget = Maths::Clamp(zoomTarget, 0.5f, 2.0f);
+}
+
+ConfigFile@ openCinematicConfig()
```
Once you read the value from the file save it to a blob/rules or global in the script and don't try to read from the file anymore.

Harrison-Miller commented on Rules/CommonScripts/CinematicCommon.as:
```diff
@@ -0,0 +1,303 @@
+	return cfg;
+}
+
+void setCinematicEnabled(bool enabled)
```
When we read, prefer reading from the value set above if we've already loaded the config.

Harrison-Miller commented on Rules/CommonScripts/PlayerCamera.as:
```diff
@@ -197,11 +203,94 @@ void onRender(CRules@ this)
+
+void onTick(CRules@ this)
+{
+	if (isCinematic())
```
this is opening and reading a file every tick, we shouldn't be doing that. We should only read the file once. And only write to it on changes.

Harrison-Miller commented on Rules/CommonScripts/PlayerCamera.as:
```diff
@@ -114,14 +116,12 @@ void onPlayerDie(CRules@ this, CPlayer@ victim, CPlayer@ attacker, u8 customData
 			else
 			{
 				camera.setTarget(null);
+				setCinematicEnabled(isCinematicEnabled());
```
Every line like this seems unnecessary if you've stored the variable correctly.

Harrison-Miller commented on Rules/CommonScripts/CinematicCommon.as:
```diff
@@ -0,0 +1,303 @@
+
+void calculateZoomTarget(float distX, float distH)
+{
+	float zoomW = calculateZoomLevelW(distX * 1.8f);
```
calculateZoomLevelW calls calculateZoomLevelH. Can we just combine this into a single function that returns a Vec2 instead of double calling calculateZoomLevelH?

eps0003 commented:
It should be ready now @Harrison-Miller
